### ISSUE TITLE
Issues with 87-based QtWebEngine (5.15.3)

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#6147]
First test results with QtWebEngine on 95e93830e45f5f10397248a794672ff6a681c8b8 (Mon Feb 8 13:34:19 2021) below.

**Comment by The-Compiler** [comment]
# "Duplicate id found" warnings for WebRTC



```
09:58:49.576 DEBUG    webview    tabbedbrowser:_on_title_changed:757 Changing title for idx 2 to 'localhost:34271/data/crashers/webrtc.html'
09:58:49.621 DEBUG    webview    tabbedbrowser:_on_title_changed:757 Changing title for idx 2 to 'Crashes on refresh'
INVALID: [887612:14:0210/095849.628886:WARNING:used_ids.h(55)] Duplicate id found. Reassigning from 104 to 127
INVALID: [887612:14:0210/095849.629372:WARNING:used_ids.h(55)] Duplicate id found. Reassigning from 106 to 125
INVALID: [887612:14:0210/095849.629635:WARNING:used_ids.h(55)] Duplicate id found. Reassigning from 110 to 124
INVALID: [887612:14:0210/095849.629828:WARNING:used_ids.h(55)] Duplicate id found. Reassigning from 112 to 123
INVALID: [887612:14:0210/095849.630061:WARNING:used_ids.h(55)] Duplicate id found. Reassigning from 102 to 122
INVALID: [887612:14:0210/095849.630420:WARNING:used_ids.h(55)] Duplicate id found. Reassigning from 103 to 121
INVALID: [887612:14:0210/095849.630569:WARNING:used_ids.h(55)] Duplicate id found. Reassigning from 105 to 120
INVALID: [887612:14:0210/095849.630830:WARNING:used_ids.h(55)] Duplicate id found. Reassigning from 111 to 119
INVALID: [887612:14:0210/095849.630976:WARNING:used_ids.h(55)] Duplicate id found. Reassigning from 109 to 118
```

```
ERROR tests/end2end/features/test_misc_bdd.py::test_webrtc_renderer_process_crash - end2end.fixtures.testprocess.InvalidLine:
```

Can probably just be ignored.

**Comment by The-Compiler** [comment]
# test_webenginedownloads data URL workaround

Looks like `download.pdf` changed to `qwe_download.pdf` in some places and the workaround isn't needed anymore.

```
FAILED tests/unit/browser/webengine/test_webenginedownloads.py::TestDataUrlWorkaround::test_workaround[True] - AssertionError: assert equals failed
FAILED tests/unit/browser/webengine/test_webenginedownloads.py::TestDataUrlWorkaround::test_workaround[False] - AssertionError: assert equals failed
FAILED tests/unit/browser/webengine/test_webenginedownloads.py::TestDataUrlWorkaround::test_workaround_needed[True]
FAILED tests/unit/browser/webengine/test_webenginedownloads.py::TestDataUrlWorkaround::test_workaround_needed[False]
```

```

**Comment by The-Compiler** [comment]
# test_new_chromium / darkmode

```
    def test_new_chromium():
        """Fail if we encounter an unknown Chromium version.
    
        Dark mode in Chromium (or rather, the underlying Blink) is being changed with
        almost every Chromium release.
    
        Make this test fail deliberately with newer Chromium versions, so that
        we can test whether dark mode still works manually, and adjust if not.
        """
>       assert version._chromium_version() in [
            'unavailable',  # QtWebKit
            '61.0.3163.140',  # Qt 5.10
            '65.0.3325.230',  # Qt 5.11
            '69.0.3497.128',  # Qt 5.12
            '73.0.3683.105',  # Qt 5.13
            '77.0.3865.129',  # Qt 5.14
            '80.0.3987.163',  # Qt 5.15.0
            '83.0.4103.122',  # Qt 5.15.2
        ]
E       AssertionError: assert '87.0.4280.144' in ['unavailable', '61.0.3163.140', '65.0.3325.230', '69.0.3497.128', '73.0.3683.105', '77.0.3865.129', ...]
E        +  where '87.0.4280.144' = <function _chromium_version at 0x7fcdd0e7e430>()
E        +    where <function _chromium_version at 0x7fcdd0e7e430> = version._chromium_version

tests/unit/browser/webengine/test_darkmode.py:236: AssertionError
```

```
FAILED tests/unit/browser/webengine/test_darkmode.py::test_new_chromium - AssertionError: assert '87.0.4280.144' in ['unavailable', '61.0.3163.140', '65.0.3325.230', '69.0.3497.128', '73.0.3683.105', '77.0.3865.129', ...]
```

**Comment by The-Compiler** [comment]
# test_preferred_colorscheme

```
        quteproc_new.send_cmd(':jseval matchMedia("(prefers-color-scheme: dark)").matches')
>       quteproc_new.wait_for(message='True')
E       end2end.fixtures.testprocess.WaitForTimeout: Timed out after 5000ms waiting for {'message': 'True'}.
```

```
10:00:08.925 DEBUG    commands   command:run:551 Calling qutebrowser.browser.commands.CommandDispatcher.jseval(<qutebrowser.browser.commands.CommandDispatcher>, 'matchMedia("(prefers-color-scheme: dark)").matches', False, False, False, world=None)
----> found it

----> Waiting for 'True' in the log
10:00:08.939 INFO     message    message:info:82 False
```

```
FAILED tests/end2end/test_invocations.py::test_preferred_colorscheme - end2end.fixtures.testprocess.WaitForTimeout: Timed out after 5000ms waiting for {'message': 'True'}.
```

On dev branch as well apparently?

**Comment by The-Compiler** [comment]
# qutescheme bdd tests

```
----> Waiting for re.compile("(load status for <qutebrowser\\.browser\\..* tab_id=\\d+ url='qute://help/quickstart\\.… in the log
09:59:11.469 DEBUG    misc       app:on_focus_object_changed:583 Focus object changed: <PyQt5.QtWidgets.QWidget object at 0x7f56a062bd30>
09:59:11.480 DEBUG    webview    browsertab:_on_navigation_request:1052 navigation request: url qute://help/img/cheatsheet-big.png, type Type.link_clicked, is_main_frame True
09:59:11.483 DEBUG    webview    webenginetab:_store_match_data:148 Active search match: 0/0
09:59:11.483 DEBUG    webview    browsertab:_set_load_status:994 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=9 url='qute://help/'>: LoadStatus.loading
09:59:11.483 DEBUG    signals    signalfilter:_filter_signals:87 emitting: cur_load_status_changed(<LoadStatus.loading: 6>) (tab 0)
09:59:11.484 DEBUG    signals    signalfilter:_filter_signals:87 emitting: cur_load_started() (tab 0)
09:59:11.499 DEBUG    network    webenginequtescheme:requestStarted:95 Got request for qute://help/img/cheatsheet-big.png
09:59:11.499 DEBUG    misc       qutescheme:data_for_url:140 url: qute://help/img/cheatsheet-big.png, path: /img/cheatsheet-big.png, host help
09:59:11.499 DEBUG    network    webenginequtescheme:requestStarted:119 Returning image/png data
09:59:11.559 DEBUG    signals    signalfilter:_filter_signals:87 emitting: cur_url_changed(PyQt5.QtCore.QUrl('qute://help/img/cheatsheet-big.png')) (tab 0)
09:59:11.560 DEBUG    webview    tabbedbrowser:_on_title_changed:757 Changing title for idx 0 to 'qute://help/img/cheatsheet-big.png'
09:59:12.886 DEBUG    webview    tabbedbrowser:_on_title_changed:757 Changing title for idx 0 to 'cheatsheet-big.png (3342×2060)'
09:59:13.181 DEBUG    webview    browsertab:_set_load_status:994 load status for <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=9 url='qute://help/img/cheatsheet-big.png'>: LoadStatus.success
09:59:13.181 DEBUG    signals    signalfilter:_filter_signals:87 emitting: cur_load_status_changed(<LoadStatus.success: 2>) (tab 0)
09:59:13.184 DEBUG    sessions   sessions:save:334 Saving session _autosave to /tmp/qutebrowser-basedir-8_clvu13/data/sessions/_autosave.yml...
09:59:13.186 DEBUG    signals    signalfilter:_filter_signals:87 emitting: cur_load_finished(True) (tab 0)
09:59:13.186 DEBUG    modes      modeman:leave:427 Ignoring leave request for KeyMode.insert (reason load started) as we're in mode KeyMode.normal
09:59:13.187 DEBUG    modes      tabbedbrowser:_leave_modes_on_load:737 Ignoring leave_on_load request due to setting.
```

```
FAILED tests/end2end/features/test_qutescheme_bdd.py::test_opening_link_with_qutehelp - end2end.fixtures.testprocess.WaitForTimeout: Timed out while waiting for qute://help/quickstart.html to be loaded
FAILED tests/end2end/features/test_qutescheme_bdd.py::test_opening_a_link_with_qutehelp - end2end.fixtures.testprocess.WaitForTimeout: Timed out while waiting for qute://help/quickstart.html to be loaded
```

Clicks a different link than previously? Might be present on master as well maybe?

**Comment by The-Compiler** [comment]
# abort in ref_counted.h

```
[886987:891645:0210/100042.022108:FATAL:ref_counted.h(76)] Check failed: CalledOnValidSequence(). 
#0 0x7fcdc887d253 base::debug::CollectStackTrace()
#1 0x7fcdc86a616d base::debug::StackTrace::StackTrace()
#2 0x7fcdc86a6124 base::debug::StackTrace::StackTrace()
#3 0x7fcdc8682fb2 logging::LogMessage::~LogMessage()
#4 0x7fcdc868371c logging::LogMessage::~LogMessage()
#5 0x7fcdc868b087 logging::CheckError::~CheckError()
#6 0x7fcdc1ffcf76 base::subtle::RefCountedBase::Release()
#7 0x7fcdc2146f68 base::RefCounted<>::Release()
#8 0x7fcdc214ae24 scoped_refptr<>::Release()
#9 0x7fcdc2149989 scoped_refptr<>::~scoped_refptr()
#10 0x7fcdda409db7 __run_exit_handlers
#11 0x7fcdda409f5e __GI_exit
#12 0x7fcdd343e892 _XDefaultIOError
#13 0x7fcdb29140be ioErrorHandler()
#14 0x7fcdd343eb9f _XIOError
#15 0x7fcdd343bbcc (/usr/lib/libX11.so.6.4.0+0x40bcb)
#16 0x7fcdd343c4d5 _XFlush
#17 0x7fcdd343f1f6 _XGetRequest
#18 0x7fcdb1d25d8b (/usr/lib/libGLX_mesa.so.0.0.0+0x37d8a)
#19 0x7fcdc9471792 gl::GLXApiBase::glXDestroyContextFn()
#20 0x7fcdc947fd80 gl::GLContextGLX::Destroy()
#21 0x7fcdc94806b2 gl::GLContextGLX::~GLContextGLX()
#22 0x7fcdc94806da gl::GLContextGLX::~GLContextGLX()
#23 0x7fcdc200c5cd base::RefCounted<>::DeleteInternal<>()
#24 0x7fcdc2009094 base::DefaultRefCountedTraits<>::Destruct()
#25 0x7fcdc20059e2 base::RefCounted<>::Release()
#26 0x7fcdc20a9490 scoped_refptr<>::Release()
#27 0x7fcdc20a93d5 scoped_refptr<>::~scoped_refptr()
#28 0x7fcdc944a7fb scoped_refptr<>::reset()
#29 0x7fcdc9449bd6 _ZN13scoped_refptrIN2gl9GLContextEEaSEDn
#30 0x7fcdcad981ed gpu::gles2::GLES2DecoderImpl::Destroy()
#31 0x7fcdcb0b9c80 gpu::CommandBufferStub::Destroy()
#32 0x7fcdcb0b8291 gpu::CommandBufferStub::~CommandBufferStub()
#33 0x7fcdcb0bc765 gpu::GLES2CommandBufferStub::~GLES2CommandBufferStub()
#34 0x7fcdcb0bc7a8 gpu::GLES2CommandBufferStub::~GLES2CommandBufferStub()
#35 0x7fcdcb0cf6da std::default_delete<>::operator()()
#36 0x7fcdcb0cb568 std::unique_ptr<>::~unique_ptr()
#37 0x7fcdcb0c3d84 gpu::GpuChannel::OnDestroyCommandBuffer()
#38 0x7fcdcb0d28d5 _ZN4base20DispatchToMethodImplIPN3gpu10GpuChannelEMS2_FviESt5tupleIJiEES6_IJEEJLm0EEJEEEvRKT_T0_OT1_PT2_St16integer_sequenceImJXspT3_EEESH_ImJXspT4_EEE
#39 0x7fcdcb0cf3dc base::DispatchToMethod<>()
#40 0x7fcdcb0cb110 IPC::MessageT<>::Dispatch<>()
#41 0x7fcdcb0c27e5 gpu::GpuChannel::OnControlMessageReceived()
#42 0x7fcdcb0c2e03 gpu::GpuChannel::HandleMessageHelper()
#43 0x7fcdcb0c2eed gpu::GpuChannel::HandleOutOfOrderMessage()
#44 0x7fcdcb0d7599 base::internal::FunctorTraits<>::Invoke<>()
#45 0x7fcdcb0d4f35 base::internal::InvokeHelper<>::MakeItSo<>()
#46 0x7fcdcb0d1959 _ZN4base8internal7InvokerINS0_9BindStateIMN3gpu10GpuChannelEFvRKN3IPC7MessageEEJNS_7WeakPtrIS4_EES6_EEEFvvEE7RunImplISA_St5tupleIJSC_S6_EEJLm0ELm1EEEEvOT_OT0_St16integer_sequenceImJXspT1_EEE
#47 0x7fcdcb0ce0ec base::internal::Invoker<>::RunOnce()
#48 0x7fcdc201fbfc _ZNO4base12OnceCallbackIFvvEE3RunEv
#49 0x7fcdc879f037 base::TaskAnnotator::RunTask()
#50 0x7fcdc87e97a6 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl()
#51 0x7fcdc87e8fc8 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork()
#52 0x7fcdc86eb1dd base::MessagePumpDefault::Run()
#53 0x7fcdc87e9fd7 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run()
#54 0x7fcdc8756f80 base::RunLoop::Run()
#55 0x7fcdc8842dab base::Thread::Run()
#56 0x7fcdc8843327 base::Thread::ThreadMain()
#57 0x7fcdc88a3f4f base::(anonymous namespace)::ThreadFunc()
#58 0x7fcdda3b13e9 start_thread
#59 0x7fcdda4ca293 __GI___clone
Task trace:
#0 0x7fcdcb0c0fc4 gpu::GpuChannelMessageFilter::OnMessageReceived()
#1 0x7fcdc96b23f8 mojo::SimpleWatcher::Context::Notify()
#2 0x7fcdc3af398c gpu::GpuChannelHost::Send()
IPC message handler context: 0xD60BEA4C

Fatal Python error: Aborted

Thread 0x00007fcdda252740 (most recent call first):
<no Python frame>
[1]    886987 abort (core dumped)  PYTHONPATH=. .venv-pyqt5153/bin/pytest --qute-bdd-webengine -- 
```

Might need to retry with a non-debug build. Maybe related to the WebRTC IDs thing?

**Comment by The-Compiler** [comment]
# Issues with dark mode and prefers-color-scheme: dark

```
>       assert content == expected_text
E       AssertionError: assert equals failed
E         'Light preference detected.'  'Dark preference detected.'
```

```
FAILED tests/end2end/test_invocations.py::test_preferred_colorscheme_with_dark_mode - AssertionError: assert equals failed
```

**Comment by The-Compiler** [comment]
# Issues with `TestWebEngineVersions::test_real_chromium_version`

```
    def test_real_chromium_version(self, qapp):
        """Compare the inferred Chromium version with the real one."""
        try:
            from PyQt5.QtWebEngine import PYQT_WEBENGINE_VERSION_STR
        except ImportError as e:
            # QtWebKit or QtWebEngine < 5.13
            pytest.skip(str(e))
    
        from qutebrowser.browser.webengine import webenginesettings
        webenginesettings.init_user_agent()
        expected = webenginesettings.parsed_user_agent.upstream_browser_version
    
        versions = version.WebEngineVersions.from_pyqt(PYQT_WEBENGINE_VERSION_STR)
>       assert versions.chromium == expected
E       AssertionError: assert equals failed
E         '83.0.4103.122'  '87.0.4280.144'
```

```
FAILED tests/unit/utils/test_version.py::TestWebEngineVersions::test_real_chromium_version - AssertionError: assert equals failed
```

Should probably skip it or something on dev versions?

**Comment by The-Compiler** [comment]
Note to self, here's the [changelog](https://codereview.qt-project.org/c/qt/qtwebengine/+/335435) for 5.15.3:

<details>

```
Qt 5.15.3 is a bug-fix release. It maintains both forward and backward
compatibility (source and binary) with Qt 5.15.2.

For more details, refer to the online documentation included in this
distribution. The documentation is also available online:

  https://doc.qt.io/qt-5.15/index.html

The Qt version 5.15 series is binary compatible with the 5.14.x series.
Applications compiled for 5.14 will continue to run with 5.15.

Some of the changes listed in this file include issue tracking numbers
corresponding to tasks in the Qt Bug Tracker:

  https://bugreports.qt.io/

Each of these identifiers can be entered in the bug tracker to obtain more
information about a particular change.

****************************************************************************
*                          Library                                         *
****************************************************************************

General
-------
  - node.js is now a hard build-time dependency
  - [QTBUG-72368] Fixed crash on macOS if the filesystem was set to being
    case-sensitive.
  - [QTBUG-85731] Fixed screen sharing on Google Meet by supporting the
    Chrome hangout extension.
  - [QTBUG-86034] Fixed blank popups in QML on EGLFS.
  - [QTBUG-86286] Fixed regression with set[Url]RequestInterceptor
    replacing existing interceptor.
  - [QTBUG-86389] Fixed issues with multi touch input.
  - [QTBUG-86406] Added support for running in Rosetta2 on ARM macs.
  - [QTBUG-87089] Fixed various issues with loadFinished signal.
  - [QTBUG-88116] Fixed building against system ICU 68.
  - [QTBUG-88708] Fixed building with MSVC 16.8.x
  - [QTBUG-88861] Fixed custom headers set on QWebEngineUrlRequestInfo
    when redirecting.
  - [QTBUG-88976] Fixed regression in font subsetting in printed PDFs.
  - [QTBUG-89001] Fixed event.getModifierState("CapsLock")
  - [QTBUG-89408] Fixed flaky crash on Windows debug builds.
  - [QTBUG-89740] Fixed crash on sites trying to use the InstalledApp
    Chrome feature. This included linkedin.com.
  - [QTBUG-89753] prefer-color-scheme now works again.
  - [QTBUG-90035] More features working in the builtin pdf-viewer, including
    zooming and localization.
  - [QTBUG-90355] Fixed suggested filename for data: URLs.
  - Added close button to devtools (for when embedded).
  - More chrome WebUIs made accessible (chrome://tracing, chrome://webrtc-logs,
    chrome://user-actions)

QtPdf
-----
  - [QTBUG-87154] Static builds of can now reuse bundled third-party
    libraries from qtbase.
  - [QTBUG-88938] Fixed loading local files on Windows.


Chromium
--------

  - The Chromium version has been updated to 87.0.4280.144
  - Security fixes from Chromium up to version 88.0.4324.150, including:
    - CVE-2020-16044: Use after free in WebRTC
    - CVE-2021-21118: Insufficient data validation in V8
    - CVE-2021-21119: Use after free in Media
    - CVE-2021-21120: Use after free in WebSQL
    - CVE-2021-21121: Use after free in Omnibox
    - CVE-2021-21122: Use after free in Blink
    - CVE-2021-21123, CVE-2021-21125, CVE-2021-21129,CVE-2021-21130, CVE-2021-21131, CVE-2021-21141:
        Insufficient data validation in File System API
    - CVE-2021-21126: Insufficient policy enforcement in extensions
    - CVE-2021-21127: Insufficient policy enforcement in extensions
    - CVE-2021-21128: Heap buffer overflow in Blink
    - CVE-2021-21132: Inappropriate implementation in DevTools
    - CVE-2021-21135: Inappropriate implementation in Performance API
    - CVE-2021-21137: Inappropriate implementation in DevTools
    - CVE-2021-21140: Uninitialized Use in USB
    - CVE-2021-21145: Use after free in Fonts
    - CVE-2021-21146: Use after free in Navigation
    - CVE-2021-21147: Inappropriate implementation in Skia
    - CVE-2021-21148: Heap buffer overflow in V8
    - CVE-2021-21149: Stack overflow in Data Transfer
    - CVE-2021-21150: Use after free in Downloads
    - CVE-2021-21152: Heap buffer overflow in Media
    - CVE-2021-21153: Stack overflow in GPU Process
    - CVE-2021-21156: Heap buffer overflow in V8
    - CVE-2021-21157: Use after free in Web Sockets
    - Security bug 937131
    - Security bug 1097499
    - Security bug 1127774
    - Security bug 1135594
    - Security bug 1144646
    - Security bug 1161654
    - Security bug 1162198
    - Security bug 1171954
    - WebRTC bug 12105
```

</details>
